### PR TITLE
extend slicing of tensors

### PR DESCRIPTION
**Pull request by vieting**
Extend slicing of tensors to support slicing with start, stop and end on tensors with on or multiple dimensions.

### CODE REVIEW

File: pytorch_to_returnn/torch/nn/modules/operator.py [hunk 1]
```diff
@@ -176,6 +176,25 @@ def create_returnn_layer_dict(self, input: Tensor) -> Dict[str, Any]:
       "position": self.pos}
 
 
+class Slice(Module):
+  """
+  Slicing of tensors. Wraps RETURNN's SliceLayer.
+  """
+  is_original_torch_module = False
+
+  def __init__(self, axis, start=None, stop=None, step=None):
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/operator.py:185]: Add type info.
```suggestion
  def __init__(self, axis: int, start: Optional[int] = None, stop: Optional[int] = None, step: Optional[int] = None):
```

File: pytorch_to_returnn/torch/nn/modules/operator.py [hunk 2]
```diff
@@ -176,6 +176,25 @@ def create_returnn_layer_dict(self, input: Tensor) -> Dict[str, Any]:
...
+    super(Slice, self).__init__()
+    self.axis = axis
+    self.start = start
+    self.stop = stop
+    self.step = step
+
+  def create_returnn_layer_dict(self, input: Tensor) -> Dict[str, Any]:
+    return {
+      "class": "slice", "axis": self.axis, "slice_start": self.start, "slice_end": self.stop, "slice_step": self.step,
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/operator.py:194]: Axis is wrong. Use `_get_returnn_axis_from_input` (or so -- see other layers).

File: pytorch_to_returnn/torch/tensor.py [hunk 3]
```diff
@@ -122,10 +122,24 @@ def float(self):
...
+    if isinstance(item, int):
+      from .nn import Gather
+      return Gather(dim=0, pos=item)(self)
+    elif isinstance(item, slice):
+      from .nn import Slice
+      return Slice(axis=0, start=item.start, stop=item.stop, step=item.step)(self)
+    elif isinstance(item, tuple):
+      assert len(item) == self.ndim
+      from .nn import Slice
+      out = None
```
Comment (albertz) [inline pytorch_to_returnn/torch/tensor.py:135]:
```suggestion
      out = self
```

File: pytorch_to_returnn/torch/tensor.py [hunk 4]
```diff
@@ -122,10 +122,24 @@ def float(self):
...
+      for ax, ax_slice in enumerate(item):
+        assert isinstance(ax_slice, slice)
+        out = Slice(axis=ax, start=ax_slice.start, stop=ax_slice.stop, step=ax_slice.step)(out or self)
```
Comment (albertz) [inline pytorch_to_returnn/torch/tensor.py:138]:
```suggestion
        out = Slice(axis=ax, start=ax_slice.start, stop=ax_slice.stop, step=ax_slice.step)(out)
```
(Always try to write code as simple as possible, and as least error-prone as possible, with as least branching (`if`, `or`, ...) as possible.)

Comment (albertz) [reply pytorch_to_returnn/torch/tensor.py:138]: You can make one other optimization though:
```
if not ax_slice.start and not ax_slice.stop and ax_slice.step in {1, None}:
  continue
```
This will look nicer in the resulting RETURNN net.

File: tests/test_layers.py [hunk 5]
```diff
@@ -241,6 +241,39 @@ def model_func(wrapped_import, inputs: torch.Tensor):
...
+
+def test_slice_2d():
+  n_batch, n_time = 3, 7
+
+  def model_func(wrapped_import, inputs: torch.Tensor):
+    if typing.TYPE_CHECKING or not wrapped_import:
+      import torch
+    else:
+      torch = wrapped_import("torch")
+    x = inputs[:, 3:]
```
Comment (albertz) [inline tests/test_layers.py:269]:
```suggestion
    x = inputs[:, 3:]
```

File: tests/test_layers.py [hunk 6]
```diff
@@ -241,6 +241,39 @@ def model_func(wrapped_import, inputs: torch.Tensor):
...
+def test_slice_1d():
+  n_batch, n_time = 3, 7
+
+  def model_func(wrapped_import, inputs: torch.Tensor):
+    if typing.TYPE_CHECKING or not wrapped_import:
+      import torch
+    else:
+      torch = wrapped_import("torch")
+    x = inputs[0]  # (T)
+    x = x[:3]
```
Comment (albertz) [inline tests/test_layers.py:253]:
```suggestion
    x = inputs[0]  # (T)
    x = x[:3]
```

File: pytorch_to_returnn/torch/tensor.py (hunk 4 above)
Comment (vieting) [reply pytorch_to_returnn/torch/tensor.py:138]: Sure, I'll do that. But I don't think this will happen often, will it?

Comment (albertz) [reply pytorch_to_returnn/torch/tensor.py]: This happens exactly in your test case. I think this is quite common.

Comment (vieting) [reply pytorch_to_returnn/torch/tensor.py]: Oh yes, it's the single colon. I see.
